docs/design/mscclpp-dsl-json-file.md

```diff
@@ -0,0 +1,478 @@
+- ```channels```: A list of communication channels, where each element describes a channel.
+- ```bufferCollection```: A list with all the remote buffers used on the algorithm.
+
+### Channels
+The channel field describes the characteristics of the channels. Basically, we have three types of channels:
+
+- Memory Channel: A MemoryChannel wraps data transfer methods that use thread-copy mechanism, ie., directly use GPU threads for writing to peer GPUs memory.
+- Port Channel: A PortChannel implements primitives when data transfer is done over ports connected to GPU memory, such as cudaMemcpy for intra-node DMA or ibv_post_send for RDMA. 
+- Switch Channel: A SwitchChannel provides primitives for performing collective operations among GPUs. These operations usually require specialized hardware sup-
+port.
+
+The Memory Channel has the following fields:
+- ```type```: Specifies the type of the channel, which in the case of the Memory Channel, will be ```memory```.
+- ```connectedTo```: Specifies the connections between channels. For example, if we have a list like: [1, 2, 3], it indicates that there are three channels: the first is connected to rank 1, the second to rank 2, and the third to rank 3.
+
+Example:
+
+```json
+"channels": [
+  {
+    "type": "memory",
+    "connectedTo": [
+      1,
+      2,
+      3
+    ]
+  }
+]
+```
+
+The Port Channel has the following fields:
+- ```type```: Specifies the type of the channel, which in the case of the Memory Channel, will be ```port```.
+- ```connectedTo```: Specifies the connections between channels. For example, if we have a list like: [1, 2, 3], it indicates that there are three channels: the first is connected to rank 1, the second to rank 2, and the third to rank 3.
+
+Example:
+
+```json
+"channels": [
+  {
+    "type": "port",
+    "connectedTo": [
+      1,
+      2,
+      3
+    ]
+  }
+]
+```
+
+The Switch Channel has the following fields:
+- ```type```: Specifies the type of the channel, which in the case of the Memory Channel, will be ```switch```.
+- ```buff```: Consist of the buffer type which the Switch Channel will be binded, this could have the following values: "i" for the input buffer, "o" for the output buffer, "s" for the scratch buffer.
+- ```rankGroups```: Consist of the group of the ranks connected by the Switch Channel, this field contains the size and the list of the ranks connected.
+
+Example:
+
+```json
+"channels": [
+  {
+    "type": "switch",
+    "buff": "i",
+    "rankGroups": [
+      {
+        "size": 8,
+        "ranks": [
+          0,
+          1,
+          2,
+          3,
+          4,
+          5,
+          6,
+          7
+        ]
+      }
+    ]
+  }
+]
+```
+
+### Buffer Collection
+The buffer collection field will contain the description of all the remote buffers that his GPU need to access, it will contain the following fields:
+
+-
+-
+-
+
+Example
+
+```json
+"bufferCollection": [
+  {
+    "rank": 1,
+    "buff": "i"
+  }
```

Comment:
Do we need some info for this buffer is for intra node or inter node. Or the buffer can be bond to memory channel or proxy channel. For proxy channel we need to call `ProxyService::addMemory(RegisteredMemory memory)` to register the buffer. for memory channel we need to keep buffer remote addr, it's different